pkg/services/rpcsrv/server.go

```diff
@@ -3075,3 +3077,60 @@ func (s *Server) Addresses() []string {
 	}
 	return res
 }
+
+func (s *Server) getRawNotaryPool(_ params.Params) (any, *neorpc.Error) {
+	if !s.chain.P2PSigExtensionsEnabled() {
+		return nil, neorpc.NewInternalServerError("P2PSignatureExtensions are disabled")
+	}
+	nrp := s.coreServer.GetNotaryPool()
+	res := &result.RawNotaryPool{Hashes: make(map[util.Uint256][]util.Uint256)}
+	nrp.IterateVerifiedTransactions(func(tx *transaction.Transaction, data any) bool {
+		if data != nil {
+			d := data.(*payload.P2PNotaryRequest)
+			mainHash := d.MainTransaction.Hash()
+			fallbackHash := d.FallbackTransaction.Hash()
+			res.Hashes[mainHash] = append(res.Hashes[mainHash], fallbackHash)
+		}
+		return true
+	})
+	return res, nil
+}
+
+func (s *Server) getRawNotaryTransaction(reqParams params.Params) (any, *neorpc.Error) {
+	if !s.chain.P2PSigExtensionsEnabled() {
+		return nil, neorpc.NewInternalServerError("P2PSignatureExtensions are disabled")
+	}
+
+	txHash, err := reqParams.Value(0).GetUint256()
+	if err != nil {
+		return nil, neorpc.ErrInvalidParams
+	}
+	nrp := s.coreServer.GetNotaryPool()
+	// Try to find fallback transaction.
+	tx, ok := nrp.TryGetValue(txHash)
+	if !ok {
+		var mainTx *transaction.Transaction
+		cont := true
+		// Try to find main transaction.
+		nrp.IterateVerifiedTransactions(func(tx *transaction.Transaction, data any) bool {
+			if data != nil {
+				mainTx = data.(*payload.P2PNotaryRequest).MainTransaction
```

Comment:
You need to make a copy, otherwise we may end up in race when Notary service is trying to modify main transaction witnesses, and RPC server is trying to serialize it.

Reply:
I'd suggest to do something like:
```
...
if data.(*payload.P2PNotaryRequest).MainTransaction.Hash().Equals(txHash) {
      // copy main tx...
      return false
}
...
```

Reply:
@tatiana-nspcc, here's the line where Notary service changes the transaction: https://github.com/nspcc-dev/neo-go/blob/b4dff7b0401ebda362789f8810245a9c52547fa5/pkg/services/notary/notary.go#L405 

That's why you need to make a copy of the main transaction by dereferencing and then referencing the structure. After that update the set of copy's witnesses.

Reply:
And here's another example:
https://github.com/nspcc-dev/neo-go/blob/b4dff7b0401ebda362789f8810245a9c52547fa5/pkg/services/notary/notary.go#L323

Reply:
OK, we have disagreement, and I was thinking of something like:
```
package main

import (
	"fmt"

	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
	"github.com/nspcc-dev/neo-go/pkg/util/slice"
)

func main() {
	tx := &transaction.Transaction{
		Scripts: []transaction.Witness{
			{
				InvocationScript: []byte{1, 2, 3},
			},
		},
	}
	cp := *tx
	txCp := &cp
	txCp.Scripts = make([]transaction.Witness, len(tx.Scripts))
	for i := range txCp.Scripts {
		txCp.Scripts[i] = transaction.Witness{
			InvocationScript:   slice.Copy(tx.Scripts[i].InvocationScript),
			VerificationScript: slice.Copy(tx.Scripts[i].VerificationScript),
		}
	}
	tx.Scripts[0].InvocationScript[1] = 8
	tx.Nonce = 3
	fmt.Println(tx.Scripts[0].InvocationScript)
	fmt.Println(txCp.Scripts[0].InvocationScript)
	fmt.Println(tx.Nonce)
	fmt.Println(txCp.Nonce)
}
```

Another possible solution would be to perform encoding-decoding operation to create transaction copy and use it as return value from RPC handler.

@roman-khimov, what do you think, do we need this at all?

Reply:
OK, actually we create a copy of mempooled transaction by Notary service:
https://github.com/nspcc-dev/neo-go/blob/b4dff7b0401ebda362789f8810245a9c52547fa5/pkg/services/notary/notary.go#L265-L274. Let me check whether we do the same for the fallback.

Reply:
We don't create the copy of fallback, and it was OK before we add these two RPCs. I'll create a fix in a separate PR.
https://github.com/nspcc-dev/neo-go/blob/b4dff7b0401ebda362789f8810245a9c52547fa5/pkg/services/notary/notary.go#L279-L281

Reply:
https://github.com/nspcc-dev/neo-go/pull/3108 resolves this discussion, @tatiana-nspcc, now you don't need to create a copy here.

Reply:
Great! Thank you very much for your help.